executor/hnswpsql.py

```diff
@@ -156,34 +171,50 @@ def _sync(
         batch_size: int = 100,
         **kwargs,
     ):
-        if timestamp is None:
-            if rebuild:
-                timestamp = datetime.min
-            elif self._vec_indexer.last_timestamp:
-                timestamp = self._vec_indexer.last_timestamp
+        # we prevent race conditions with search
+        with self.lock:
+            if timestamp is None:
+                if rebuild:
+                    # we assume all db timestamps are UTC +00
+                    timestamp = datetime.fromtimestamp(0, timezone.utc)
+                elif self._vec_indexer.last_timestamp:
+                    timestamp = self._vec_indexer.last_timestamp
+                else:
+                    self.logger.error(
+                        f'No timestamp provided in parameters: '
+                        f'and vec_indexer.last_timestamp'
+                        f'was None. Cannot do sync'
+                    )
+                    return
             else:
-                self.logger.error(
-                    f'No timestamp provided in parameters: '
-                    f'and vec_indexer.last_timestamp'
-                    f'was None. Cannot do sync'
-                )
-                return
-        else:
-            timestamp = datetime.fromisoformat(timestamp)
+                timestamp = datetime.fromisoformat(timestamp)
 
-        iterator = self._kv_indexer._get_delta(
-            shard_id=self.runtime_args.pea_id,
-            total_shards=self.total_shards,
-            timestamp=timestamp,
-        )
+            iterator = self._kv_indexer._get_delta(
+                shard_id=self.runtime_args.pea_id,
+                total_shards=self.total_shards,
+                timestamp=timestamp,
+            )
 
-        if rebuild or self._vec_indexer.size == 0:
-            # call with just indexing
-            self._vec_indexer = HnswlibSearcher(**self._init_kwargs)
-            self._vec_indexer.index_sync(iterator, batch_size)
+            if rebuild or self._vec_indexer.size == 0:
```

Comment:
u can move vec index lock aquiring here.

And maybe have another lock for the kv indexer

Reply:
Yes, that's what I did. Can you check again?